internal/rbd/replicationcontrollerserver.go

```diff
@@ -0,0 +1,349 @@
+		return librbd.ImageMirrorModeSnapshot, nil
+	}
+
+	var mirroringMode librbd.ImageMirrorMode
+	switch imageMirroringMode(val) {
+	case imageMirrorModeSnapshot:
+		mirroringMode = librbd.ImageMirrorModeSnapshot
+	case imageMirrorModeJournal:
+		mirroringMode = librbd.ImageMirrorModeJournal
+	default:
+		return mirroringMode, status.Errorf(codes.InvalidArgument, "%s %s not supported", imageMirroringKey, val)
+	}
+	return mirroringMode, nil
+}
+
+// cleanup performs below resource cleanup operations.
+func (cs *ControllerServer) cleanup(rbdVol *rbdVolume) {
+	if rbdVol != nil {
+		// destroy the cluster connection
+		rbdVol.Destroy()
+		// release the volume lock
+		cs.VolumeLocks.Release(rbdVol.VolID)
+	}
+	if cs.CR != nil {
+		// remove the keyfile
+		cs.CR.DeleteCredentials()
+	}
+}
+
+// EnableVolumeReplication enables replication on volume.
+func (cs *ControllerServer) EnableVolumeReplication(ctx context.Context,
+	req *replication.EnableVolumeReplicationRequest,
+) (*replication.EnableVolumeReplicationResponse, error) {
+	var rbdVol = &rbdVolume{}
+	var err error
+	// ger rbdVolume from VolumeID
+	rbdVol, err = cs.getVolumeFromID(ctx, req.GetVolumeId(), req.GetSecrets())
+	defer cs.cleanup(rbdVol)
+	if err != nil {
+		return nil, err
+	}
+	// extract the mirroring mode
+	mirroringMode, err := getMirroringMode(ctx, req.GetParameters())
+	if err != nil {
+		return nil, err
+	}
+
+	mirroringInfo, err := rbdVol.getImageMirroringInfo(ctx)
+	if err != nil {
+		err = fmt.Errorf("failed to get mirroring info of %s with error %s", rbdVol.String(), err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mirroringInfo.State != librbd.MirrorImageEnabled {
+		err = rbdVol.enableImageMirroring(ctx, mirroringMode)
+		if err != nil {
+			util.ErrorLog(ctx, "failed to enable mirroring of %s with error %s", err.Error())
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+	return &replication.EnableVolumeReplicationResponse{}, nil
+}
+
+// DisableVolumeReplication disables replication on volume.
+func (cs *ControllerServer) DisableVolumeReplication(ctx context.Context,
+	req *replication.DisableVolumeReplicationRequest,
+) (*replication.DisableVolumeReplicationResponse, error) {
+	var rbdVol = &rbdVolume{}
+	var err error
+	// ger rbdVolume from VolumeID
+	rbdVol, err = cs.getVolumeFromID(ctx, req.GetVolumeId(), req.GetSecrets())
+	defer cs.cleanup(rbdVol)
+	if err != nil {
+		return nil, err
+	}
+
+	// extract the force option
+	force, err := getForceOption(ctx, req.GetParameters())
+	if err != nil {
+		return nil, err
+	}
+
+	mirroringInfo, err := rbdVol.getImageMirroringInfo(ctx)
+	if err != nil {
+		err = fmt.Errorf("failed to get mirroring info of %s with error %s", rbdVol.String(), err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mirroringInfo.State == librbd.MirrorImageEnabled {
+		err = rbdVol.disableImageMirroring(ctx, force)
```

Comment:
Unsure if image is secondary if mirroring can be disabled, may require an additional check  in that regard.

Reply:
possible to disable with force option from CLI. added a check now

Reply:
From a use case perspective, I am thinking something like so,

- Provide ability to delete volume when it is secondary, by actually not deleting anything (as the mirroring would recreate the same), IOW fake the result as success as long as image is secondary
- This prevents access via VolRep CR as PV is not longer available (as well as PVC)
- Hence, DisableMirroring would also act to see if it is secondary and if so ignore the request instead of forcing it

The other way to go is to force disable, the assumption (that we need to test) is that the mirroring daemon does not bring the image back due to primary existing and active elsewhere. (also curious how this state is maintained on the secondary)